### PR TITLE
Add lint rule for requiring useEffect deps

### PR DESCRIPTION
**Pull request by ahuth**
[sc-224374]

This PR adds a lint rule for requiring `useEffect` deps.

I'm 50/50 on whether we should introduce this or not, because
- There are valid reasons for not having a deps array (e.g. we want to run something after every render/update)
- But I think in our case it's often an oversight or copypasta (and if it was intentional we can safely suppress the lint rule)

Let me know your thoughts 🙏

### CODE REVIEW

File: packages/eslint-plugin-edu-react/package.json [hunk 1]
```diff
@@ -1,23 +1,23 @@
 {
-  "name": "@chanzuckerberg/eslint-config-edu-react",
+  "name": "@chanzuckerberg/eslint-plugin-edu-react",
```
Comment (ahuth) [inline packages/eslint-plugin-edu-react/package.json:2]: To add a custom rule I had to convert this from a "config" to a "plugin" (eslint is kind of weird, and "configs" are just config, but "plugins" can have config and new rules).

An alternative would be to separate into both a config and a plugin, but that seems like too much work and not enough benefit to me.

File: packages/eslint-plugin-edu-react/src/index.ts [hunk 2]
```diff
@@ -7,6 +13,7 @@ module.exports = {
     },
   },
   rules: {
+    '@chanzuckerberg/eslint-plugin-edu-react/use-effect-deps-presence': 'error',
```
Comment (ahuth) [inline packages/eslint-plugin-edu-react/src/index.ts:16]: Maybe this should be

```suggestion
    '@chanzuckerberg/eslint-plugin-edu-react/use-effect-deps-presence': 'warning',
```

🤔

**Comment by booc0mtaco** [comment]
I think I would be in favor of it being a warning to start. Assuming the main penalty is potential performance impacts, a warning could notify individuals about this, and permit those deps-array-less scenarios without an ignore comment. It's also a good code review topic (e.g., "can you explain why no deps array is needed here")

Are there any other worse outcomes I'm ignoring?

**Comment by dierat** [comment]
>There are valid reasons for not having a deps array (e.g. we want to run something after every render/update)

I think that in the case something does need to run again on every update,
- that's rare enough that it's not that big of a deal to add a lint override comment
- we would probably want someone to leave a comment explaining it anyway

I'm fine with the lint rule, and I'm thinking of it as a reminder to consider how you actually want your code to run.

**Comment by ahuth** [comment]
@booc0mtaco

> Are there any other worse outcomes I'm ignoring?

Not really. Worst case the logic is different than intended, but that's a danger for any code 😂 . Although it may be particularly easy to forget the deps array and get different logic.

**Comment by booc0mtaco** [comment]
@ahuth ah gotcha. I bias conservative on introducing restrictions, so after some more thought, I change to keeping introducing it as error, as it seems not to be too bad to be more strict. We could always loosen it later if it does cause annoyance

**Comment by timzchang** [comment]
I think some of this has already been said, so to add/retell a few thoughts:

Some potential tradeoffs of adding this rule:
✅ - protects against people adding useEffect without a dependency array unintentionally. In >95%(?) of cases, one would want there to be a dependency array.
❌ - it's not traditional to have an error for something that's not technically wrong, and people might be confused

I don't really think the con is something to be that concerned about since the error message is very clear and actionable. So overall feel good about this change!

**Review by palasha — COMMENTED** [review]
What does this add that `react-hooks/exhaustive-deps` doesnt?

**Comment by ahuth** [comment]
> What does this add that react-hooks/exhaustive-deps doesnt?

Given a deps array, `exhaustive-deps` makes sure it has all required entries in it.

But it doesn't ensure that there's a deps array in the first place.

File: packages/eslint-plugin-edu-react/src/index.ts [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
     react: {
       version: 'detect',
     },
   },
   rules: {
-    'jsx-a11y/anchor-ambiguous-text': 'error',
+    // prettier-ignore
+    '@chanzuckerberg/edu-react/use-effect-deps-presence': 'error',
+    // TODO: Turn this rule on when it's released. It's not present in v6.6.1.
+    'jsx-a11y/anchor-ambiguous-text': 'off',
```
Comment (ahuth) [inline packages/eslint-plugin-edu-react/src/index.ts:19]: While testing this I discovered that this rule isn't actually released, yet 😅

File: packages/eslint-plugin-edu-react/src/index.ts [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
+  plugins: ['@chanzuckerberg/edu-react', 'jsx-a11y', 'react', 'react-hooks'],
   settings: {
     react: {
       version: 'detect',
     },
   },
   rules: {
-    'jsx-a11y/anchor-ambiguous-text': 'error',
+    // prettier-ignore
+    '@chanzuckerberg/edu-react/use-effect-deps-presence': 'error',
```
Comment (ahuth) [inline packages/eslint-plugin-edu-react/src/index.ts:17]: Weirdly the "full" name (`eslint-plugin-edu-react` vs `edu-react`) doesn't work. No idea why (docs indicate it should).

**Pull request merged**